src/ui/public/accessibility/kbn_accessible_click.js

```
@@ -0,0 +1,86 @@
+.directive('kbnAccessibleClick', function ($parse) {
+  return {
+    restrict: 'A',
+    scope: {
+      kbnAccessibleClick: '&',
+    },
+    controller: $element => {
+      $element.on('keydown', e => {
+        // We don't care if the user is actually interacting with a different element.
+        if (e.currentTarget !== e.target) {
+          return;
+        }
+
+        // Prevent a scroll from occurring if the user has hit space.
+        if (accessibleClickKeys[e.keyCode]) {
+          e.preventDefault();
+        }
+      });
+    },
+    link: (scope, element, attrs) => {
+      // The whole point of this directive is to hack in functionality that native buttons provide
+      // by default.
+      const elementType = element.prop('tagName');
+
+      if (elementType === 'BUTTON') {
+        throw new Error(`kbnAccessibleClick doesn't need to be used a button.`);
+      }
+
+      if (elementType === 'A' && attrs.href !== undefined) {
+        throw new Error(`kbnAccessibleClick doesn't need to be used on a link if it has a href attribute.`);
+      }
+
+      // If the developer hasn't already specified attributes required for accessibility, add them.
+      if (attrs.tabindex === undefined) {
+        element.attr('tabindex', '0');
+      }
+
+      if (attrs.role === undefined) {
+        element.attr('role', 'button');
+      }
+
+      element.on('keyup', e => {
+        // Support keyboard accessibility by toggling display on ENTER or SPACE keypress.
+        if (accessibleClickKeys[e.keyCode]) {
+          // If a unique click handler has been specified, then call it.
+          if (attrs.kbnAccessibleClick) {
+            return scope.kbnAccessibleClick();
+          }
+
+          // Otherwise, default to the specified ng-click.
+          const ngClick = $parse(attrs.ngClick);
+
+          if (!ngClick) {
+            throw new Error('kbnAccessibleClick requires a callback or ng-click defined on its element.');
+          }
```

Could you move this out of the event handler and into the main link function so that it fails early if ngClick is not specified?